### PR TITLE
Refactor and fix clipboard export, refs #13395

### PR DESCRIPTION
**Pull request by djjuhasz**
- Refactor clipboard export code to reduce duplication of code
- Fix coding  standard issues such as long lines, and trailing whitespace
- Add conditional to avoid trying to export external digital objects, and add notice to log
- Fix bug that always set the job name to "XML export" even for CSV exports
- Add more comments

### CODE REVIEW

File: apps/qubit/modules/clipboard/actions/exportAction.class.php [hunk 1]
```diff
@@ -3,18 +3,18 @@
...
  *
  * Access to Memory (AtoM) is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
- * You should have received a copy of the GNU General Public License
- * along with Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
+ * You should have received a copy of the GNU General Public License along with
+ * Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
```
Comment (jraddaoui) [inline apps/qubit/modules/clipboard/actions/exportAction.class.php]: This is going to be needed in **a lot** of files. We still need to decide the standard we want to follow (re: #13317) but both, PSR-2 and PSR-12, state that _"There MUST NOT be a hard limit on line length; the soft limit MUST be 120 characters; lines SHOULD be 80 characters or less"_.

https://www.php-fig.org/psr/psr-12/#23-lines

So maybe we should not worry that much about it, specially in here.

Comment (djjuhasz) [reply apps/qubit/modules/clipboard/actions/exportAction.class.php]: Yes, you have a good point.  The thing that galls me about the current statement is that it takes just as many lines when it's wrapped at 80 characters. 😜 

I find that PSR statement on line lengths to be so confusing and ambiguous - but I agree that it allows for the current statement as is. 😕

File: apps/qubit/modules/clipboard/actions/exportAction.class.php [hunk 2]
```diff
@@ -432,30 +433,20 @@ protected function addField($name)
...
+              'It is not possible to select both digital objects and'
+              . ' descendants for export at the same time. Digital objects can'
+              . ' only be exported for records that are on the clipboard.'
+            );
           }
+
+          $this->helpMessages[] = __(
+            'Digital objects with restricted access or copyright will not'
+            . ' be exported.'
+          );
```
Comment (jraddaoui) [inline apps/qubit/modules/clipboard/actions/exportAction.class.php:445]: This should not appear for repositories.

File: lib/job/arActorExportJob.class.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+ * (at your option) any later version.
+ *
+ * Access to Memory (AtoM) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
+ */
```
Comment (jraddaoui) [inline lib/job/arActorExportJob.class.php:18]: For example ... :stuck_out_tongue:

If you really want to change this, maybe do a search and replace over the entire AtoM folder in a different commit.

File: lib/job/arActorExportJob.class.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+   */
+  protected function doExport($path)
+  {
+    $search = self::findExportRecords($this->params);
+
+    // Scroll through results then iterate through resulting IDs
+    foreach (
+      arElasticSearchPluginUtil::getScrolledSearchResultIdentifiers($search)
+      as $id
+    ) {
```
Comment (jraddaoui) [inline lib/job/arActorExportJob.class.php]: New line for the bracket. This is one case I'd prefer in one line even if it's higher than 80 chars.

File: lib/job/arExportJob.class.php [hunk 5]
```diff
@@ -229,126 +241,104 @@ protected function addDigitalObjects(&$zip, &$errors)
...
-  protected function getDigitalObjects()
+  protected function addDigitalObject($resource, $tempDir)
   {
-    // Prepare array for digital object ids
-    $digitalObjects = array();
+    // Skip if the includeDigitalObjects option is not set, or it is disabled
+    if (
+      !isset($this->params['includeDigitalObjects'])
+      || !$this->params['includeDigitalObjects']
+    ) {
```
Comment (jraddaoui) [inline lib/job/arExportJob.class.php]: New line for bracket.

File: lib/job/arExportJob.class.php [hunk 6]
```diff
@@ -229,126 +241,104 @@ protected function addDigitalObjects(&$zip, &$errors)
...
+          && $this->user->isAuthenticated()
+          && QubitAcl::check($resource, 'read')
+        ) || (
+          'informationObject' == $this->params['objectType']
+          && QubitAcl::check($resource, 'readMaster')
+          && QubitGrantedRight::checkPremis($resource->id, 'readMaster')
+          && !$digitalObject->hasConditionalCopyright()
+        )
+      )
+    ) {
```
Comment (jraddaoui) [inline lib/job/arExportJob.class.php]: New line for bracket.

File: lib/job/arExportJob.class.php [hunk 7]
```diff
@@ -229,126 +241,104 @@ protected function addDigitalObjects(&$zip, &$errors)
...
+      // export is allowed
```
Comment (jraddaoui) [inline lib/job/arExportJob.class.php]: Uppercase first.

File: lib/job/arInformationObjectCsvExportJob.class.php [hunk 8]
```diff
@@ -23,232 +23,82 @@
...
-          $job->info($job->i18n->__(
-            '%1 items exported.', array('%1' => $itemsExported))
-          );
-        }
-
-        $itemsExported++;
+    // Export descendants if option was selected
+    if (!$this->params['current-level-only'])
+    {
+      foreach ($resource->getDescendantsForExport($options) as $item) {
```
Comment (jraddaoui) [inline lib/job/arInformationObjectCsvExportJob.class.php]: New line for bracket.

File: lib/job/arInformationObjectExportJob.class.php [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+   * @return arElasticSearchPluginQuery  AtoM Elasticsearch query
+   */
+  static public function getCurrentArchivalStandard()
+  {
+    // If not using RAD, default to ISAD CSV export format
+    $archivalStandard = 'isad';
+    if (
+      QubitSetting::getByNameAndScope('informationobject', 'default_template')
+      == 'rad'
+    ) {
```
Comment (jraddaoui) [inline lib/job/arInformationObjectExportJob.class.php]: New line for bracket. And another one I'd prefer in a single line.

File: lib/job/arInformationObjectExportJob.class.php [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    if (0 == $search->count())
+    {
+      return;
+    }
+
+    // Scroll through results then iterate through resulting IDs
+    foreach (
+      arElasticSearchPluginUtil::getScrolledSearchResultIdentifiers($search)
+      as $id
+    ) {
```
Comment (jraddaoui) [inline lib/job/arInformationObjectExportJob.class.php]: New line for bracket. And the same.

File: lib/job/arExportJob.class.php [hunk 11]
```diff
@@ -255,18 +258,34 @@ protected function addDigitalObject($resource, $tempDir)
 
     $digitalObject = $resource->getDigitalObject();
 
+    // Skip this digital object if it doesn't exist in the database, is a URI
+    // for an external object, or export is not authorized
     if (
       empty($digitalObject)
       || !$this->allowDigitalObjectExport($resource, $digitalObject)
     ) {
```
Comment (jraddaoui) [inline lib/job/arExportJob.class.php]: Not sure if I missed this one.

File: lib/job/arExportJob.class.php [hunk 12]
```diff
@@ -255,18 +258,34 @@ protected function addDigitalObject($resource, $tempDir)
...
       return false;
     }
 
+    // Don't try to export an external digital object
+    if (QubitTerm::EXTERNAL_URI_ID == $digitalObject->usageId)
```
Comment (jraddaoui) [inline lib/job/arExportJob.class.php]: I think we may want to check OFFLINE_ID too (I think it comes only from the metadata DIP upload from AM).

**Review by jraddaoui — CHANGES_REQUESTED** [review]
Looks great @djjuhasz, thanks!

Only a couple of important comments, all the other cosmetics.

File: lib/job/arActorExportJob.class.php (hunk 4 above)
Comment (djjuhasz) [reply lib/job/arActorExportJob.class.php]: Fixed

File: lib/job/arExportJob.class.php (hunk 5 above)
Comment (djjuhasz) [reply lib/job/arExportJob.class.php]: I like the `) {` convention (obviously 😜), but I think you are right that putting open bracket on a new line is canonical. 😢

Comment (djjuhasz) [reply lib/job/arExportJob.class.php]: Fixed

File: apps/qubit/modules/clipboard/actions/exportAction.class.php (hunk 1 above)
Comment (djjuhasz) [reply apps/qubit/modules/clipboard/actions/exportAction.class.php]: Reverted

File: apps/qubit/modules/clipboard/actions/exportAction.class.php (hunk 2 above)
Comment (djjuhasz) [reply apps/qubit/modules/clipboard/actions/exportAction.class.php:445]: It doesn't, the `$this->digitalObjectsAvailable` flag above is false for the repository type.

File: lib/job/arInformationObjectCsvExportJob.class.php (hunk 8 above)
Comment (djjuhasz) [reply lib/job/arInformationObjectCsvExportJob.class.php]: Fixed

File: lib/job/arInformationObjectExportJob.class.php (hunk 9 above)
Comment (djjuhasz) [reply lib/job/arInformationObjectExportJob.class.php]: Done. 😄

File: lib/job/arInformationObjectExportJob.class.php (hunk 10 above)
Comment (djjuhasz) [reply lib/job/arInformationObjectExportJob.class.php]: Done.

**Review by djjuhasz — COMMENTED** [review]
Feedback addressed :)

File: lib/job/arExportJob.class.php (hunk 12 above)
Comment (djjuhasz) [reply lib/job/arExportJob.class.php]: Oops, missed this one in my commit :(

File: lib/job/arExportJob.class.php (hunk 11 above)
Comment (djjuhasz) [reply lib/job/arExportJob.class.php]: Do you mean it doesn't check if it's an external URI like the comment indicates, or the `) {`?

Comment (jraddaoui) [reply lib/job/arExportJob.class.php]: Yeah, the bracket.

**Review by jraddaoui — APPROVED** [review]
Thanks @djjuhasz!

Just a few not addressed comments, the only one that may be important is to check the OFFLINE_ID usage type.

File: lib/job/arExportJob.class.php (hunk 12 above)
Comment (djjuhasz) [reply lib/job/arExportJob.class.php]: Addressed in commit 9a5ba5e3a

File: lib/job/arExportJob.class.php (hunk 7 above)
Comment (djjuhasz) [reply lib/job/arExportJob.class.php]: Okay, got this in a *third* commit 🤦

File: lib/job/arExportJob.class.php (hunk 11 above)
Comment (djjuhasz) [reply lib/job/arExportJob.class.php]: Fixed

**Comment by djjuhasz** [comment]
> Just a few not addressed comments, the only one that may be important is to check the OFFLINE_ID usage type.

Sorry, missed this the first time I went through your comments. :(  I've addressed it now.

**Review by jraddaoui — APPROVED** [review]
:+1:

**Pull request merged**